### ISSUE TITLE
Not Handling Connections to Multiple Worker Processes Correctly

### ISSUE THREAD

**Issue by BoogeyOff** [justpy-org/justpy#208]
if you just change the example `qnotify.py` script (or indeed any other) to run on two worker processes as follows:
```
jp.justpy(notify_test, start_server=False)
app = jp.app    

if __name__ == '__main__':
    import uvicorn
    uvicorn.run('qnotify:app',host='127.0.0.1',port=80,workers=2,use_colors=False)
```
keep refreshing the page and hitting one of the arrow buttons. Eventually the buttons will stop working and you will get the error message on every button click:
`WARNING justpy: No page to load`
After a Ctrl+C in the stack trace you will see:
```
  File "c:\users\dima\anaconda3\lib\site-packages\justpy\justpy.py", line 279, in on_disconnect
    await WebPage.instances[pid].on_disconnect(websocket)   # Run the specific page disconnect function
KeyError: 0
```
I believe this stack trace actually points to the issue where the page stops working, even though you don't normally get to see it when it occurs. I've only come to this conclusion after running the script via `pm2`, which opens separate command line windows for each forked worker process.

Whenever the workers load balance a page load, so that the page http request goes to one worker and the websocket connection to the other (easy to see in separate shell windows), that's when the exception occurs and the page stops working.

Maybe it's an implementation issue with Uvicorn, but it is justpy that's throwing the exception.

A refresh will usually still get the page back.

Btw, the KeyError continues to increment if the same issue re-occurs.

**Comment by BoogeyOff** [comment]
Btw, exact same issue when running `JustPy` on `HyperCorn` with multiple workers:

`hypercorn -b 127.0.0.1:80 --workers 2 qnotify:app`

**Comment by elimintz** [comment]
Thanks for flagging this. I'll take a look.

**Comment by platinops** [comment]
Have you been able to take a look at this? I'm having the same issue, even with the most basic hello world script, both with `hypercorn` and `uvicorn` (on Windows) - when using multiple workers in combination with websockets.

Without websockets, all seems to run fine, but then I my app (and any other?) loses functionality of pushing data from server to client.

**Comment by elimintz** [comment]
As I am not familiar with the details of the worker implementation, perhaps you can help me with this. A justpy app is an extension of a starlette app which sits on top of uvicorn. When there are two workers, does that mean that there are two identical apps running in parallel?  If that is the case, they will not have access to each other's web pages and components. And when an event on a webpage created by one worker is sent to the other worker, the results are unpredictable and will cause errors.

**Comment by platinops** [comment]
Yes, I think that is the case. Uvicorn will spawn three processes in this case; one managing process and two worker processes. Each of those workers will contain its own app. Incoming requests will then be distributed among the workers depending on how busy they are.
This is explained quite well [here](https://fastapi.tiangolo.com/deployment/server-workers/#uvicorn-with-workers).

**Comment by elimintz** [comment]
How do I change justpy to work well with multiple workers? What directions should I explore? Any suggestions welcome.

**Comment by platinops** [comment]
That's a good question - I am by no means an expert on this :)

The issue with the current setup seems to be the one described in the "Tip" at the bottom of [this FastAPI page](https://fastapi.tiangolo.com/advanced/websockets/).

As I understand it from reading below links, you would need a caching mechanism (preferably in a local database?). All workers would publish their incoming websocket data to the cache, and they would also subscribe to it. That way they would effectively share incoming websocket data between each other and each worker extracts the info it needs in order to update their own app/pages.
- https://github.com/encode/starlette/issues/216
- https://github.com/encode/starlette/issues/133
- https://github.com/taoufik07/nejma-chat

On the above-linked `fastapi` page, a link is made to the encode team's [broadcaster](https://github.com/encode/broadcaster) package. All out-of-memory (and thus multiprocessing-proof) caching mechanisms seem to be beyond the scope of Justpy however, using Redis/Kafka/Postgres. 

Also touching upon the issue of making an app 'stateless': https://dash.plotly.com/sharing-data-between-callbacks

**Comment by kaihasse** [comment]
Coincidentally, I also did some research on this topic today...
Maybe [this thread](https://stackoverflow.com/questions/65686318/sharing-python-objects-across-multiple-workers) about sharing data across multiple FastAPI workers is helpful as well...

**Comment by elimintz** [comment]
Thank you for the suggestions and the info. Will study the issue further. There does not seem to be a simple fix.

**Comment by BoogeyOff** [comment]
Hey guys, the work-around for this issue that I've been using is to have NGINX as the external-facing server, acting as a load-balancer/reverse proxy for a number of back-end single-worker JustPy instances. It works quite well and I daresay this is a more realistic setup for a scalable system. I'm sure there are many similar alternatives.

A number of advantages to a setup like this:
 - let nginx serve the static content
 - ability to direct all /templates/ request to a separate JustPy instances purely dedicated for that purpose. That way your actual dynamic app can be scaled separately
 - let nginx handle all http to https redirections
 - let nginx restrict acceptable ssl cipher suites
 - weighted load balancing to reflect back-end server specs

If you wanna pursue this route, be sue to specify **ip_hash** as the load balance mechanism. That way your web socket connections will be forwarded to the same JustPy instance that served the initial http request. NGINX also needs a few extra lines of config to forward socket connections, doesn't do it right out of the box.

Happy to provide an example if anyone else is interested in going down that route.

**Comment by elimintz** [comment]
Thanks, this is very helpful. I will mention this in the documentation

**Comment by platinops** [comment]
@BoogeyOff Thanks for the feedback! Yes, I would definitely be interested in an example showing the setup you describe.

**Comment by BoogeyOff** [comment]
No problems. Below is a minimal example of the config you'd need for nginx. Then you'd run JustPy instances on ports 10001 and 10002 over https. You can give it the same key and crt files as for nginx.

The bulk of the SSL config probably isn't "necessary", but you might as well follow what according to my research is close to best practice. I've omitted a lot of other things I mentioned above, eg nginx serving static files - there's about a million config settings for nginx and it would just confuse the issue to post a full config here.

Feel free to play around with this. My memory is a bit hazy on what some of the settings do exactly.

	events {
		worker_connections 1024;
	}

	http {

	        # to prevent leaking info on your server and system - test on https://gf.dev/server-signature-test
	        server_tokens off;

		# think this is for web sockets (can't remember)
		map $http_upgrade $connection_upgrade {
			default upgrade;
			'' close;
		}
		
		# the end-points for dynamic content servers to load balance to
		upstream webapp_dyn {
			ip_hash;
			server 127.0.0.1:10001 weight=1; # avoid using "localhost" here
			server 127.0.0.1:10002 weight=1;
		}

		# the actual https server config (redirects to backend to according to webapp_dyn config above)
		server {
		
			listen 443 http2 ssl;
			listen [::]:443 http2 ssl;

			server_name your.web.domain;
			
			location / {
				# hook into the back-end servers
				proxy_pass https://webapp_dyn; # note https
				proxy_redirect off;
				proxy_buffering off;
				proxy_connect_timeout 10s;
				client_max_body_size 20M; # upload size limit
				
				# turn off the cache for the dynamic content
				proxy_no_cache 1;
				# even if cached, don't try to use it
				proxy_cache_bypass 1; 
				# another way to kill the cache
				expires -1;
				
				# this section is for websockets I think
				proxy_http_version 1.1;
				proxy_set_header Upgrade $http_upgrade;
				proxy_set_header Connection $connection_upgrade;
				proxy_set_header Host $host;


			}

			# --------------------- the ssl security config -----------------------------------
			
			# give it the site certificate and key
			# note: with proper certificate chain, just give it one bundle file that's a cat of the authorities and the site
			# generate via: cat your_domain.crt intermediate.crt root.crt >> ssl-bundle.crt
			# see https://www.ssls.com/knowledgebase/how-to-install-an-ssl-certificate-on-a-nginx-server/
			
			ssl_certificate your_site.crt;
			ssl_certificate_key your_site.key;

			# disallow 1 and 1.1, re TLS security see https://www.packetlabs.net/tls-1-1-no-longer-secure/
			ssl_protocols TLSv1.3 TLSv1.2;
			
			ssl_prefer_server_ciphers on;
			ssl_session_timeout 1d;
			ssl_session_cache shared:SSL:50m;
			ssl_session_tickets off; # Requires nginx >= 1.5.9
			ssl_stapling on; # Requires nginx >= 1.3.7
			ssl_stapling_verify on; # Requires nginx => 1.3.7  
			add_header Strict-Transport-Security "max-age=15768000; includeSubDomains; preload";
			
			# use ONLY the strongest ciphers! from https://syslink.pl/cipherlist/
			ssl_ciphers EECDH+AESGCM:EDH+AESGCM;
			ssl_ecdh_curve secp384r1; 				# Requires nginx >= 1.1.0

			# diffie hellman key exchange params
			ssl_dhparam .\dhparam.pem; # generate file via: openssl dhparam -out .\dhparam.pem 4096

		}

	}



	#

**Comment by platinops** [comment]
This was very helpful, much appreciated!

Only tested it with two identical app instances running on different ports, and works like a charm. Both sections you mentioned were indeed required for websockets to work.

My config below (on nginx-1.21.6). Note this is running on an internal network, not facing the www.

```
worker_processes  1;

events {
    worker_connections  1024;
}

http {

	# To prevent leaking info on your server and system - test on https://gf.dev/server-signature-test
	server_tokens off;

	# For websockets
	map $http_upgrade $connection_upgrade {
		default upgrade;
		'' close;
	}

	# The end-points to load balance to
	upstream webapp {
		
		# Make sure that requests from each IP are handled by the same worker
		ip_hash;
		
		# This directive is not compatible with ip_hash and messes up websockets => do not use
		# least_conn; 
		
		# Workers - each one running a single Uvicorn>Starlette>Justpy app instance 
		# Avoid using "localhost" here
		server 127.0.0.1:10001; 
		server 127.0.0.1:10002;
	}

    server {

		listen 443 http2 ssl;
		listen [::]:443 http2 ssl;
		
		location / {
		
			# Hook into the back-end servers
			proxy_pass https://webapp; # note https
			proxy_redirect off;
			proxy_buffering off;
			proxy_connect_timeout 10s;
			client_max_body_size 20M; # upload size limit
			
			# This section is for websockets
			proxy_http_version 1.1;
			proxy_set_header Upgrade $http_upgrade;
			proxy_set_header Connection $connection_upgrade;
			proxy_set_header Host $host;

		}
		
		# SSL config
		
		ssl_certificate      ssl/my_site.crt;
		ssl_certificate_key  ssl/my_site.key;
		
		# Disallow 1 and 1.1, re TLS security see https://www.packetlabs.net/tls-1-1-no-longer-secure/
		ssl_protocols TLSv1.3 TLSv1.2;
		
		ssl_prefer_server_ciphers on;
		ssl_session_timeout 1d;
		ssl_session_cache shared:SSL:50m;
		ssl_session_tickets off;   # Requires nginx >= 1.5.9
		ssl_stapling on;           # Requires nginx >= 1.3.7
		ssl_stapling_verify on;    # Requires nginx => 1.3.7  
		add_header Strict-Transport-Security "max-age=15768000; includeSubDomains; preload";
		
		# Use ONLY the strongest ciphers! from https://syslink.pl/cipherlist/
		ssl_ciphers EECDH+AESGCM:EDH+AESGCM;
		ssl_ecdh_curve secp384r1;  # Requires nginx >= 1.1.0
		
    }
	
}
```

**Comment by WolfgangFahl** [comment]
Won't fix - recommend to use nicegui if better fastapi handling is needed.
